### ISSUE TITLE
Issue returning struct with implementations

### ISSUE THREAD

**Issue by Zaitam** [fzyzcjy/flutter_rust_bridge#585]
While using the new feature added in the PR https://github.com/fzyzcjy/flutter_rust_bridge/pull/543, I found an issue while returning an struct containg an implementation.
Ex code:
```rs 
pub fn get_sum_struct() -> SumWith {
    SumWith {x: 21}
}

pub struct SumWith { pub x: u32 }

impl SumWith {
    pub fn sum(&self, y: u32) -> u32 { self.x + y }
}
```
> Highlight on the first function, `get_sum_struct`

If we now run the flutter_rust_bridge, no errors will be found. Anyhow, there is an issue in the `bridge_generated.dart` file.

Dart bridge generated code:
```dart
Future<SumWith> getSumStruct({dynamic hint}) => executeNormal(FlutterRustBridgeTask(
        callFfi: (port_) => inner.wire_get_sum(port_),
        parseSuccessData: _wire2api_sum_with,
        constMeta: kGetSumConstMeta,
        argValues: [],
        hint: hint,
      ));
```

The problem is with the `_wire2api_sum_with` var. It's type is of `SumWith Function(Native, dynamic)` instead of the expected `SumWith Function(dynamic)`.

I think the issue is quite simple to fix, and can be done by adding a nullable parameter, even thought I don't think they exist in rust. I haven't got the needed exprience in Rust to take hands in the matter and make a PR.

**Comment by fzyzcjy** [comment]
Hmm sounds reasonable.

/cc @lattice0 who implemented this wonderful feature.

**Comment by lattice0** [comment]
Fix: https://github.com/fzyzcjy/flutter_rust_bridge/pull/587

**Comment by Zaitam** [comment]
So... 
To start with, it now works for singular items. Thanks @lattice0 for that.

However it doesn't work for arrays nor vectors.
```rs
pub fn get_sum_struct() -> [SumWith; 1] {
    [SumWith {x: 21}]
}
```
```rs
pub fn get_sum_struct() -> Vec<SumWith> {
    vec![SumWith {x: 21}]
}
```

```dart
List<SumWith> _wire2api_list_get_sum_struct(dynamic raw) {
  return (raw as List<dynamic>).map(_wire2api_get_sum_struct).toList();
}
```
With similar issue where `_wire2api_get_sum_struct` is of type `SumWith Function(Native, dynamic)` instead of `dynamic Function(dynamic)`

Think the issue is still on the same size, so I am not creating a new issue. If @fzyzcjy can reopen it till this is also fixed.

**Comment by lattice0** [comment]
These cases are very complicated as there's no `this` to be the bridge on the generated function:
```dart
List<SumWith> _wire2api_list_sum_with(dynamic raw) {
  return (raw as List<dynamic>).map(_wire2api_sum_with).toList();
}
```
which should be

```dart
List<SumWith> _wire2api_list_sum_with(dynamic raw) {
  return (raw as List<dynamic>).map((d)=>_wire2api_sum_with(bridge, d)).toList();
}
```

but then `_wire2api_list_sum_with` should receive `bridge` as well which would force
```dart
  Future<List<SumWith>> getSumStructArray({dynamic hint}) =>
      executeNormal(FlutterRustBridgeTask(
        callFfi: (port_) => inner.wire_get_sum_struct_array(port_),
        parseSuccessData: _wire2api_list_sum_with,
        constMeta: kGetSumStructArrayConstMeta,
        argValues: [],
        hint: hint,
      ));
```

to work with bridge as well.

Also `_wire2api_list_sum_with` doesn't even know that `_wire2api_sum_with` needs a bridge, so it's even harder

**Comment by fzyzcjy** [comment]
I am thinking about something even more general: Shall we pass an argument like `class Wire2ApiParseContext {TheBridgeClass bridge;}` (just randomly named here...) into *each and every* such wire2api functions? It is like a context so it is fine to be passed around for everyone, just like how we use BuildContext in Flutter.

Or, another way, maybe we should make those wire2api functions be *methods* inside the bridge class? Then we automatically get the bridge instance via `this`.

**Comment by lattice0** [comment]
I like the second option, but I'm out of time. I already need to finish the exception throwing, the cancelable tasks, and I'm looking for a bug in my generated code with the new master. I want to fix this though, would be nice to be able to return any composable type.

**Comment by fzyzcjy** [comment]
Sure, take your time!

@Zaitam what do you think about the proposal above? Feel free to PR, ask questions, and I will provide feedbacks, so no worries about lacking skills to fix it.

**Comment by Zaitam** [comment]
Sorry for the delayed answer.
I'll try to do a PR implementing the second idea. However I can't promise much do to lack of skill. I would like, if u can, an explanation of general terms how the integration works, due that I was unable to find it in the docs.

**Comment by fzyzcjy** [comment]
> how the integration works

Which integration?

**Comment by fzyzcjy** [comment]
Generally speaking, the `codegen` is firstly run and it generates bridge_generated.rs and bridge_generated.dart from api.rs. Then, users just normally use those rust and dart code as if they are written by hand.

**Comment by lattice0** [comment]
@Zaitam I put this at the root and run it:

```bash 
set -ex
cargo run --manifest-path=frb_codegen/Cargo.toml -- --rust-input frb_example/pure_dart/rust/src/api.rs --dart-output frb_example/pure_dart/dart/lib/bridge_generated.dart --dart-format-line-length 120
dart format --line-length 120 frb_example/pure_dart/dart/
cargo fmt
dart format --line-length 80 frb_dart
cargo build
cd ~/flutter_rust_bridge/frb_example/pure_dart/dart && dart run lib/main.dart ~/flutter_rust_bridge/target/debug/libflutter_rust_bridge_example.so
```

it compiles the project, formats everything for the CI and runs the tests on pure_dart.

A Cargo.toml with

```
[workspace]
members = ["frb_rust", "frb_example/pure_dart/rust"]
```
is also useful so rust-analyzer works.

When I started I also knew nothing about the library but with trial and error I was able to modify some stuff. Generally just printing the types with debug: `println!("type: {:?}", ty)`

**Comment by Zaitam** [comment]
I've just started reading the code but I don't understand what do you mean by the Bridge class.
> Or, another way, maybe we should make those wire2api functions be methods inside the bridge class? Then we automatically get the bridge instance via this.
Is it the generated one or another one? 

If its the first, there are already many items in the class:
```dart
abstract class FlutterRustBridgeExample {
  Future<Uint8List> drawMandelbrot(
      {required Size imageSize,
      required Point zoomPoint,
      required double scale,
      required int numThreads,
      dynamic hint});

  FlutterRustBridgeTaskConstMeta get kDrawMandelbrotConstMeta;

  Future<String> passingComplexStructs({required TreeNode root, dynamic hint});

  FlutterRustBridgeTaskConstMeta get kPassingComplexStructsConstMeta;
...
``` 
And what does this class do, in difference to `FlutterRustBridgeExample`,  `class FlutterRustBridgeExampleImpl extends FlutterRustBridgeBase<FlutterRustBridgeExampleWire>
    implements FlutterRustBridgeExample {`
    
Last, which of both should contain the `_wire2api_` functions?

Sorry for the many questions, and I still can't promise any progress, but I both require it myself, and think it would be useful

**Comment by fzyzcjy** [comment]
>  what do you mean by the Bridge class

The class in bridge_generated.dart, yes, just FlutterRustBridgeExample and its friend FlutterRustBridgeExampleImpl.

> And what does this class do, in difference to FlutterRustBridgeExample, class FlutterRustBridgeExampleImpl extends FlutterRustBridgeBase<FlutterRustBridgeExampleWire> implements FlutterRustBridgeExample {

* FlutterRustBridgeExample: abstract api definition
* FlutterRustBridgeExampleImpl: real implementation

**Comment by Zaitam** [comment]
Okay, just to get it clear:
Under the bridge of 
```rust
pub struct ConcatenateWith {
    pub a: String,
}


impl ConcatenateWith {
    pub fn test(a: String) -> ConcatenateWith {
        ConcatenateWith { a }
    }
}
```

we get generated various item in the `dart` bridge.

```dart
abstract class FlutterRustBridgeExampleSingleBlockTest {
  Future<ConcatenateWith> testStaticMethodConcatenateWith({required String a, dynamic hint});

  FlutterRustBridgeTaskConstMeta get kTestStaticMethodConcatenateWithConstMeta;
}

class ConcatenateWith {
  final FlutterRustBridgeExampleSingleBlockTest bridge;
  final String a;

  ConcatenateWith({
    required this.bridge,
    required this.a,
  });

  static Future<ConcatenateWith> test(
          {required FlutterRustBridgeExampleSingleBlockTest bridge, required String a, dynamic hint}) =>
      bridge.testStaticMethodConcatenateWith(a: a, hint: hint);
}

Future<ConcatenateWith> testStaticMethodConcatenateWith({required String a, dynamic hint}) =>
      executeNormal(FlutterRustBridgeTask(
        callFfi: (port_) => inner.wire_test__static_method__ConcatenateWith(port_, _api2wire_String(a)),
        parseSuccessData: (d) => _wire2api_concatenate_with(this, d),
        constMeta: kTestStaticMethodConcatenateWithConstMeta,
        argValues: [a],
        hint: hint,
      ));
```
(Excluding the `WireBase`, which I think is irrelevant for this, correct me if I am wrong)

Which part is the one required to be under the `class FlutterRustBridgeExampleSingleBlockTestImpl`?
If it is the last one, we just need to refactor the function `generate_dart_implementation_body` under `dart/mod.rs`. If it is another I still need to get more deep on how the project is working (Really nice work, btw).

Post this, we require to add a public var on the class containing the bridge, which I think `this` will do the job?

Finnaly, the modification is only required on the `dart` end, right? I could no read anything that seem a relevant reason why it required a mod in the `rust` part. 

Thanks for answering the previous, and wish answering this isn't harder than just solving the issue 😂

**Comment by fzyzcjy** [comment]
> Which part is the one required to be under the class FlutterRustBridgeExampleSingleBlockTestImpl?

Quote my proposal: "maybe we should make those wire2api functions be methods inside the bridge class? Then we automatically get the bridge instance via this."

Thus, it is the:

```dart
ConcatenateWith _wire2api_concatenate_with(FlutterRustBridgeExampleSingleBlockTest bridge, dynamic raw) {
  final arr = raw as List<dynamic>;
  if (arr.length != 1) throw Exception('unexpected arr length: expect 1 but see ${arr.length}');
  return ConcatenateWith(
    bridge: bridge,
    a: _wire2api_String(arr[0]),
  );
}
```

that should be in

```dart
class FlutterRustBridgeExampleSingleBlockTestImpl
```

notice it is `...Impl` class, not the declaration class.

> Finnaly, the modification is only required on the dart end, right? I could no read anything that seem a relevant reason why it required a mod in the rust part.

Guess yes. Anyway you can try starting making a PR, and see whether tests fail or not. Once all tests pass we are sure it is done :)

**Comment by fzyzcjy** [comment]
Btw https://github.com/fzyzcjy/flutter_rust_bridge/issues/593

**Comment by fzyzcjy** [comment]
bump

**Comment by Zaitam** [comment]
Fixed by #764
